src/KeyUnit.cpp

```diff
@@ -298,7 +347,88 @@ QString KeyUnit::getKeyName( int keyCode, int modifierCode )
     else return QString("undefined key");
 }
 
+void KeyUnit::initStats()
+{
+    statsKeyTotal = 0;
+    statsTempKeys = 0;
+    statsActiveKeys = 0;
+    statsActiveKeysMax = 0;
+    statsActiveKeysMin = 0;
+    statsActiveKeysAverage = 0;
+    statsTempKeysCreated = 0;
+    statsTempKeysKilled = 0;
+}
+
+void KeyUnit::_assembleReport( TKey * pChild )
+{
+    list<TKey*> * childrenList = pChild->mpMyChildrenList;
+    for( auto it2 = childrenList->begin(); it2 != childrenList->end(); it2++)
+    {
+        TKey * pT = *it2;
+        _assembleReport( pT );
+        if( pT->isActive() ) statsActiveKeys++;
+        if( pT->isTempKey() ) statsTempKeys++;
+        statsKeyTotal++;
+    }
+}
 
+QString KeyUnit::assembleReport()
+{
+    statsActiveKeys = 0;
+    statsKeyTotal = 0;
+    statsTempKeys = 0;
+    for( auto it = mKeyRootNodeList.begin(); it != mKeyRootNodeList.end(); it++)
+    {
+        TKey * pChild = *it;
+        if( pChild->isActive() ) statsActiveKeys++;
+        if( pChild->isTempKey() ) statsTempKeys++;
+        statsKeyTotal++;
+        list<TKey*> * childrenList = pChild->mpMyChildrenList;
+        for( auto it2 = childrenList->begin(); it2 != childrenList->end(); it2++)
+        {
+            TKey * pT = *it2;
+            _assembleReport( pT );
+            if( pT->isActive() ) statsActiveKeys++;
+            if( pT->isTempKey() ) statsTempKeys++;
+            statsKeyTotal++;
+        }
+    }
+    QStringList msg;
+    msg << "Keys current total: " << QString::number(statsKeyTotal) << "\n"
+        << "tempKeys current total: " << QString::number(statsTempKeys) << "\n"
+        << "active Keys: " << QString::number(statsActiveKeys) << "\n";
+        /*<< "active Aliass max this session: " << QString::number(statsActiveAliassMax) << "\n"
+        << "active Aliass min this session: " << QString::number(statsActiveAliassMin) << "\n"
+        << "active Aliass average this session: " << QString::number(statsActiveAliassAverage) << "\n"*/
+        //<< "tempAliass created this session: " << QString::number(statsTempAliassCreated) << "\n"
+        //<< "tempAliass killed this session: " << QString::number(statsTempAliassKilled) << "\n"
+        //<< "current total regex Aliass: " << QString::number(statsRegexAliass) << "\n"
+        //<< "average line processing time: " << QString::number(statsAverageLineProcessingTime) << "\n"
+        //<< "max line processing time: " << QString::number(statsMaxLineProcessingTime) << "\n"
+        //<< "min line processing time: " << QString::number(statsMinLineProcessingTime) << "\n";
+    return msg.join("");
+}
+
+void KeyUnit::markCleanup( TKey * pT )
+{
+    for( auto it = mCleanupList.begin(); it != mCleanupList.end(); it++)
+    {
+        if( *it == pT )
+        {
+            return;
+        }
+    }
+    mCleanupList.push_back( pT );
```

Comment:
:thought_balloon:  If we converted the `(std::list<TKey *>) mCleanupList` into a `(QSet<TKey *>) mCleanupSet` we would not have to look through the list each time we try to add something to check that it is not present before adding it, we have no need of the intrinsic ordering of a `std::list<T>` (or `QList<T>`) here and a "set" data structure is faster for evaluation of whether a value is a member of the container.  This method would become:

    void KeyUnit::markCleanup( TKey * pT )
    {
        mCleanupSet.insert( pT );
    }

and the following method could be slightly simplified with the Qt `foreach` container iterator:

    void KeyUnit::doCleanUp()
    {
        foreach( auto it, mCleanupSet ) {
            delete it;
        }
        mCleanupSet.clear();
    }